feat: support logout cmd

S96EA opened the pull request:
## What kind of change does this PR introduce?

feature

## What is the current behavior?

[1688](https://github.com/supabase/cli/issues/1688)

kamilogorek commented on internal/logout/logout.go:
```diff
@@ -0,0 +1,33 @@
+var loggedOutMsg = "You are now logged out."
+
+func Run(ctx context.Context, stdout *os.File, params RunParams) error {
+	err := utils.RunProgram(ctx, func(p utils.Program, ctx context.Context) error {
```
This operation should be instantaneous, I don't think we need to wrap it inside a separate program.

S96EA replied:
Yes. Thanks. I remove the "RunProgram" wrap.

kamilogorek commented on internal/logout/logout.go:
```diff
@@ -0,0 +1,33 @@
+		return err
+	}
+
+	fmt.Fprintf(stdout, "Token deleted successfully.\n\n")
+	fmt.Fprintln(stdout, loggedOutMsg)
```
```suggestion
	fmt.Fprintln(stdout, "Token deleted successfully. You are now logged out.")
```

This should be sufficient.

S96EA replied:
Thanks~

kamilogorek commented on internal/logout/logout.go:
```diff
@@ -0,0 +1,33 @@
+type RunParams struct {
+	Fsys afero.Fs
+}
+
+var loggedOutMsg = "You are now logged out."
```
```suggestion
```

S96EA replied:
Thanks~

kamilogorek commented on internal/logout/logout.go:
```diff
@@ -0,0 +1,33 @@
+var loggedOutMsg = "You are now logged out."
+
+func Run(ctx context.Context, stdout *os.File, params RunParams) error {
+	err := utils.RunProgram(ctx, func(p utils.Program, ctx context.Context) error {
```
I'd add a prompt asking users to confirm the logout saying that it will remove the token from their system.

S96EA replied:
Yes. Thanks. I added a PreRun stage to ask users.

kamilogorek commented on internal/utils/access_token.go:
```diff
@@ -84,6 +84,23 @@ func fallbackSaveToken(accessToken string, fsys afero.Fs) error {
+}
+
+func fallbackDeleteToken(fsys afero.Fs) error {
+	path, err := getAccessTokenPath()
```
It should not fail when there's no access token found. When `NotFound` error is returned, `logout` command should just print `You were not logged in, nothing to do.` or similar message. Now it'll fail with:
```
remove /Users/kamilogorek/.supabase/access-token: no such file or directory
```
When called multiple times.

S96EA replied:
Thanks. I checked if the error returned by `Remove` is ENOENT, This is OK for OS fs implementation in afero. But I didn't find a common error or check function in afero that would express different fs NotFound error.

kamilogorek commented on internal/logout/logout.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+func PreRun() error {
+	if !utils.PromptYesNo("Do you want to log out, which will remove the token from your system?", true, os.Stdin) {
```
Canceling removal should still be exitcode `0`, as it's terminating the program correctly. Not sure why project deletion is throwing error here.
However because `PreRun` cannot handle cancellation other than via error, you can move this check into `Run` and write

```go
if !utils.PromptYesNo("Do you want to log out, which will remove the token from your system?", true, os.Stdin) {
  fmt.Fprintln(stdout, "Not deleting token")
  return nil
}
```

S96EA replied:
Thanks. Done.

kamilogorek commented:
Lovely, thanks @S96EA. Added few changes myself so you don't have to go back and forth anymore :)
Please apply those to this PR and we're good to go - https://gist.github.com/kamilogorek/cf2cacf7ff0cce8b91d65734df776fad

S96EA commented:
> Lovely, thanks @S96EA. Added few changes myself so you don't have to go back and forth anymore :) Please apply those to this PR and we're good to go - https://gist.github.com/kamilogorek/cf2cacf7ff0cce8b91d65734df776fad
> 
> (changed default to `false` as its destructive operation, fixed `actual/expected` order in tests, simplified error check and used separate error value for not logged in)

I'm sincerely appreciative of your support. Thanks again.

kamilogorek commented:
LGTM! @sweatybridge feel free to merge anytime.
